### ISSUE TITLE
Optimal File Naming (Follow-up to #19)

### ISSUE THREAD

**Issue by tradej** [redhat-documentation/modular-docs#25]
@sterobin writes in #19:

> What value is achieved by naming files by module, such as proc-creating_guided_decision_tables? That is, why is it helpful to have files listed by module type?

At the risk of misinterpreting what you said in, let me present the following argument (in case I indeed misinterpreted what you said, feel free to correct me):

* You write that you do share our view that anchor IDs should not diverge from file names very much.
* You also write that you agree with our view that anchor IDs should not diverge from the module heading very much because of the URLs.

The above necessarily means that, by transition, you agree that file names should not diverge very much from the module title. From this, I infer the following:

If file names should not diverge much from module headings—which provide no meaningful sorting capability—and we need to add a piece of text to it (the module type), we can either:

* make it a suffix, not gaining any sorting capability, or
* make it a prefix, gaining a not particularly helpful, but still non-trivial sorting capability.

As a side note, I consider prepending subject matter (again, a free-form string) a considerable deviation from the module heading.

cc @rkratky @asteflova

**Comment by sterobin** [comment]
@tradej and @rkratky , you make good points. Let me sum up so we can conclude:
- Anchor IDs should be less cryptic and match module headings for clarity in URLs and in our own cross-refs.
- Anchor IDs and file names really should be the same or close, for simplicity. Thus file names should just match anchor IDs, which match module headings---i.e., module headings, anchor IDs, and file names should therefore all be the same or similar.
- A module affix in file names is helpful. As a suffix, it removes any kind of sorting for the files, only by the name of the module heading, essentially. But as a prefix, it at least allows sorting of some kind, however trivial or not, as @tradej points out.

On these grounds I suppose I will concede to the following:
- Anchor ID: **[#creating_guided_decision_tables]**
- File name: **proc_creating-guided-decision-tables.adoc**

But please humor me in one last effort to understand something. My concession is not because either of you have given a particularly strong argument about why you prefer the file sorting that results from this, as opposed to the sorting I have proposed, which is what started this discussion. It is only because of the points above and because of other elements considered, not necessarily because of a strong preference. And that's fine. @rkratky says, "I have found it a lot easier to navigate directories of modules when the modules are grouped by module type." I understand this to mean, it's easier to navigate the directories organized by modules as opposed to not being organized in any way (i.e., without module prefix). 

If you can just answer one last question, I'll be done and go with the majority rule.

Below are links to two file lists. Ignore the difference in list length. Which of these two file lists do you find easier to navigate, **and why**? 
- [File list A](http://pastebin.test.redhat.com/522041)
- [File list B](http://pastebin.test.redhat.com/522022)

Strictly for my peace of mind so I can know that we've made a deliberate decision, independent of other considerations.

**Comment by rkratky** [comment]
@sterobin, thanks for the summary. Two reasons why I find no. 2 easier to navigate:

1.  File names mirroring module headings is easier on the brain -- I don't have to 'translate' the file name to figure out what the heading is.
2.  When I'm looking for a module, I already know whether I'm looking for a proc, ref, or con -- so, I go directly to the relevant group of modules and then select from it (within a group, modules are sorted alphabetically, making orientation fast); if method no. 1 was used, I would first locate the 'topic' of the module and only then select the module's type. So, it adds a step to the process.

Btw, any reason why you prefer `proc_foo-bar.adoc` over `proc-foo_bar.adoc`? I find the latter easier on the eyes (i.e. I have no real reason, I just like to better).

**Comment by sterobin** [comment]
Thanks @rkratky . That's all I wanted to know. I'm glad you ask about underscores vs dashes. I do have a couple of reasons, actually. 

First, it's just best practice in my experience to use dashes in file names. Underscores are kind of archaic and should be avoided if possible.

That said, and second, I think there's logic in anchor names being different and using underscores. Part of the reason is that :context: attributes when topics are reused should be similar to the file name, including having dashes. And with what we're proposing, the dashes may be the only real differentiation between anchors (underscores) and files. Thus when the {context} displays the dashed attribute in URLs and in cross-refs, it's differentiated from the anchor name a bit, which I kind of like. That is, where it's "anchor_name_context-attribute". Of course, you could just give the context attribute dashes regardless of file name and achieve the same thing, but I like the attribute and file to look similar to give a sense of "this topic is being used in the context (location) of this file."

Examples: 
- "For details, see xref:creating_data_objects_defining-assets[ ]." 
- "..._User_Guide/#creating_data_objects_defining-assets"

Just my two cents.

**Comment by rkratky** [comment]
> Underscores are kind of archaic and should be avoided if possible.

This is the first time I'm hearing this. I don't think we need to go with fashion :-)

I'm afraid I don't understand your other argument. I would argue that we should use the same (i.e. either underscores or dashes) in file names, IDs, as well as context attributes. So, to use your example:

- `include::creating_data_objects_defining-{context}.adoc`
- "For details, see xref:creating_data_objects_defining-assets[ ]."
- "..._User_Guide/#creating_data_objects_defining-assets"

**Comment by sterobin** [comment]
No, it would be "creating_data_objects_{context}.adoc". The "defining-assets" part is the context attribute.

But either way, I say let's be consistent then. So long as we're going down this road, I vote for using dashes all around. Having the distinction between anchors and file names is not critical. 

Underscores are not an issue of fashion, as you suggest. It has a functional handicap. When underscores appear in URLs, they disappear and just look like spaces when the raw URL is linked and underlined, as they will be when docs start getting linked more in the product. This may sound like an edge case, but it came up again and again at my previous job. Something about how they are processed when developers add links in the product, or when users copy and paste raw links as a reference in bug tickets. etc. I was surprised at it myself, but nonetheless, we saw an improvement when we stuck to dashes.

That said, what do you think about this format then:
- Anchor ID A: **[#creating-data-objects_{context}]**
- File name A: **proc_creating-data-objects.adoc**
- Anchor ID B: **[#defining-assets_{context}]**
- File name B: **proc_defining-assets.adoc**
- Cross-ref of A+B: **See xref:creating-data-objects_defining-assets[ ].**

I think the format would still be helpful to distinguish where an anchor ends and where a context attribute begins, for our own sake. If you think that's too complicated still, then dashes all around is fine by me.

Thoughts @rkratky?

**Comment by rkratky** [comment]
> No, it would be "creating_data_objects_{context}.adoc". The "defining-assets" part is the context attribute.

OK, the misunderstanding is just in that I didn't know where the attribute started. My argument was for `creating_data_objects-{context}.adoc`.

> Underscores are not an issue of fashion, as you suggest. It has a functional handicap. When underscores appear in URLs, they disappear and just look like spaces when the raw URL is linked and underlined, as they will be when docs start getting linked more in the product.

Right, that's a valid point. I would argue that raw URLs should not be featured anywhere in the docs, but it's true the URLs will possibly be linked from user discussion forums where this would be an issue.

> I think the format would still be helpful to distinguish where an anchor ends and where a context attribute begins, for our own sake. If you think that's too complicated still, then dashes all around is fine by me.

That's exactly what I was suggesting, except underscores and dashes swapped. However, I'm happy with your way.

**Comment by sterobin** [comment]
Oh, I see. So you're saying this **xref:creating_data_objects-defining-assets** (anchor_name-{context}) shows more clearly where the context begins than this **xref:creating_data_objects_defining-assets** (anchor_name_{context})? I thought the second does, but we're probably splitting hairs. I'm fine to go with the first if you think so.

And the proposition to make everything dashes all around, in both anchors and file names, might be too large of an impact for other writing teams to accept. So maybe just shelf that for now and leave as is, dashes in files, underscores in anchors?

**Comment by rkratky** [comment]
We seem to be going in circles. One more summary. I was suggesting this:

`[#creating_data_objects-defining_assets]`

But I can live with:

`[#creating-data-objects_defining-assets]`

**Comment by sterobin** [comment]
Yes, I much prefer the second one. Dashes are better than underscores for reasons I've already given. So again, this is the new recommended format for anchor and file names, as I mentioned above:

- Anchor ID A: **[#creating-data-objects_{context}]**
- File name A: **proc_creating-data-objects.adoc**
- Anchor ID B: **[#defining-assets_{context}]**
- File name B: **proc_defining-assets.adoc**
- Cross-ref of A+B: **See xref:creating-data-objects_defining-assets[ ].**  

(Where `:context:` is set to `defining-assets` to match the file name, minus the module type prefix. In reality, it would more likely be an assembly, but another module is possible. The naming would be the same though.)

@rkratky and @tradej, are we all in agreement with this then? I'll update the PR right now in the meantime and push the change for you to see the final output and give a go no-go.

And thank you @asteflova for your comment in #19. Good advice :)

**Comment by rkratky** [comment]
+1 Thanks

**Comment by tradej** [comment]
Yes, this is acceptable for me.

**Comment by ssorj** [comment]
Hello.  I'm a passerby.  The apparent conclusion on this thread looks totally backward to me.  Underscore is a weaker separator, not a stronger one, than hyphen.  IOW, "proc_creating-data-objects" parses as "proc_creating, data, objects".  I don't think that's what you were going for.

A data point out in the world: https://www.ecreativeim.com/blog/index.php/2011/03/30/seo-basics-hyphen-or-underscore-for-seo-urls/

I recommend abandoning any notion of mixing underscores and hyphens.  People are highly confused by the difference, so it's not a good place to build a distinction.  Pick one or the other.

**Comment by bfallonf** [comment]
Hi

I don't have much to say in the great hyphens vs dashes debate, but I'd like to throw my opinion in on module type prefixes (con_, proc_, ref_, assembly_).

I'd much prefer the module types be suffixes.. So <title_con>, etc. It just opinion, and we all have one, but having all file starting with one of four prefixes is hard to look at, and even harder to read through. I'm currently converting a small doc and it's already becoming unruly...

And this might be a conversation for another thread, but I'd ask if we need module type prefix/suffixes at all. If we follow the file naming conventions in the mod docs docs[1], it's not that hard to understand to understand what kind of module it is anyway ("If it's got an -ing word, then it's a procedure!").

Are the files named this way because of the upcoming tool set? Would it make sense to have some sort of attribute inside of the file instead?

[1] https://redhat-documentation.github.io/modular-docs/#anchor-and-file-names

**Comment by mishaone** [comment]
+1 to everything Brice said. In my view, the benefit to having the same standardized title, ID and file name is to allow people not familiar or new to the documentation to quickly and easily identify what module they want to edit based on the rendered result they're probably looking at. Adding a prefix removed the alphabetized view of modules/assemblies and makes it more taxing for human eyes to find what we should be editing.

If we must have an identifier in the file name, let it be a suffix, although when you add a medium length file name and medium length human-readable context value, a suffix appended to it makes the file name incredibly long. If it's at all possible to just use metadata, perhaps something defined within a module, to indicate what kind of module it is, that would be ideal, IMO.

**Comment by vikram-redhat** [comment]
I don't care what is decided, as long as it is consistent, standardised and communicated. We are starting to modularize, but it makes it very difficult to prescribe standards to my own team if the goal posts keep moving for something as banal, yet so pervasive, as the file names. The fact that this issue is still open worries me that the standard for the file names was not a standard, as I thought it was and teams are free to do their own thing.

**Comment by dayleparker** [comment]
The topic of file name conventions has come up a few times in discussion recently, and wanted to chime in here from a writer's perspective to +1 what @bfallonf and @mishaone are suggesting. I also much prefer a suffix to indicate a file's module type, as our content will be much more time-consuming to locate a file to update it if all files are sorted by module type in a list.  A module prefix will probably result in having to run a search query to find a module (ie. in Atom), is much more time-consuming than a simple glance at an alphabetized, human-readable list. 

For example, the docs suite I work on uses images formatted as numeric names (ie. 1452.png) and it takes a bit of work to figure out which image to update (not doable via glancing at a list - you need to look at the contents of each file), and I envision a long list of "proc-procedurename.adoc" files being just as laborious to parse. For example:

![imagenames](https://user-images.githubusercontent.com/20275317/42921686-7541d87c-8b60-11e8-9b07-f73496247fbc.png)

Meanwhile, here's an example of an easy-to-skim and update list using the sidebar in Atom (where a 'con' or 'proc' on the end wouldn't affect the content author):

![topicnames](https://user-images.githubusercontent.com/20275317/42921703-8725713e-8b60-11e8-8970-1f183f9500c8.png)

Lastly, since we are actively working in the upstream and trying to get more contributors (Red Hat and otherwise) and continually attempting to break down barriers to entry, I'd prefer making content updates/creation as easy for users as possible. I really like Misha's proposal to define the module type inside the text of the file.

**Comment by tylerkelly13** [comment]
+1 to points made by @bfallonf, @dayleparker, @mishaone, & @vikram-redhat. If Red Hat is actively moving to modular documentation, we need to finalise standards as soon as possible. My own opinion reflects the preceding comments, file names and anchor IDs should be based on the heading of the module (although having context at the end of the anchor ID's is an important variation).

**Comment by laubai** [comment]
I'm in a few minds about a number of points above, so in summary:

- I think it makes sense to keep the file name closely associated with the anchor ID.
- I don't think it makes sense to place the equivalent of metadata in the file name, and module type is a kind of metadata.
- By extension, I think separating multiple kinds of content with either an underscore or a hyphen in a filename should not be necessary, so the conversation that Stetson and Robert had above about what separates words vs separating content types need not even be a necessary discussion.
- I think hyphens are easier to type than underscores, so I'd appreciate not needing to type underscores in file names as much as possible.

Basically, I think this would be much simpler if file names were just file names and we put our metadata about module type, subject matter, etc. somewhere else.

A tag or keyword taxonomy and a way to apply it to modules is necessary to resolve any of the above issues, I think.

**Comment by mishaone** [comment]
I'm actually curious about why it's important for us to know the module type at all. Surely this information is clear looking at either the rendered version or the contents of the module after matching it easily (because directly alphabetical order) with the appropriate module file in the repo.

On the other hand, if someone like a developer edits the docs, I would think that they don't really care much about the module type and have no idea what con, proc, etc. mean so there isn't much benefit to prioritising that information in such a shorthand sort of way.

With or without prefixes, suffixes, what have you, we should have some basic instructions about how to order information the modular way. I'm unsure what value is added if we remove the module type from the file name.

(From my phone so apologies for typos)

**Comment by rkratky** [comment]
@mishaone I find it a lot easier to navigate docs when I know at first glance what sort of file I'm dealing with (which is also an answer to @laubai above). I believe it also helps with collaboration: I don't have to study the contents of someone else's module to figure out what type of info it is intended to convey. Having this data point right from the start is a shortcut: "OK, I'm editing a procedure." instead of "OK, I need to guess the purpose of this module, so that I don't screw it up with my edits."

You're right about the developers though. On the other hand, devels won't have a clue about great many other things both in the repo (attributes inclusion, symlink structure, ...) and in the files themselves (template structure, formatting guidelines, ...). So, I don't think we need to be too concerned about that aspect.

**Comment by mishaone** [comment]
@rkratky that's an interesting perspective. I haven't looked at the templates recently so this may have changed but I would think the content itself would make it clear at a glance what type of module we're in, particularly so if you happen to be a writer familiar with modular docs. 

A procedure, for example, would be very difficult to mistake for anything else. I'd consider a compete rewrite if it were not clear. The same can be said for concepts and references, since one should have an introduction to a subject while the latter would be a reference-style content module.

Based on this, I'm not sure explicitly stating the module type explicitly up front (by virtue of it being the first sorting criteria) is especially useful. As I said earlier, I'm sure this metadata is useful in some way, possibly to the tooling we will eventually use, but surely that metadata can be included in a less explicit way (as good metadata should, IMO) within the module?

I haven't actually been following the mod docs updates too closely so apologies if this is a rehash of something already discussed.

**Comment by mishaone** [comment]
That's fair, @tradej.

I think the main concerns for some of us new to the discussion today were around how useful the idea of prefixes would be compared to how they disrupt the alphabetical sorting. Opinions may vary on that, of course, as we've seen.

The other concern is around the fact that as many products amp up thier efforts to adopt modular documentation, not having these things already settled means we run the risk of making a significant effort towards something that may be changed again in a few days and need more effort (or people digging in thier heels because the previous thing was what they implemented and they don't want to spend time redoing it).

**Comment by tradej** [comment]
It seems to me that the biggest pain point here is whether we want to do prefixes vs suffixes, and everything else is mostly cosmetics (I'm not saying those are not valid points, but they seem overshadowed by the former at the moment). Since it's a thing of taste, I will try move the discussion forward with this proposal:

We mandate that a repository sticks to either a prefix or a suffix, according to their preference, and consistently use that for all files in that repository. The allowed syntax is only one of the following:

    proc_doing-a-thing.adoc

or

    doing-a-thing_proc.adoc

Doing it this way doesn't break any existing tooling AFAICT, and if it does pose a problem for any tooling in the future, we can ask the teams participating to stick to either one or the other (renaming all files and modifying _include_ statements using a script, and checking the doc set is a one man-day thing, IMO).

With the implementation of this proposal, this issue would be closed.

WDYT?

**Comment by rkratky** [comment]
+1 @tradej Very nice:)

**Comment by sterobin** [comment]
+1 @tradej.

Just to throw my hat into this busy ring, this topic was discussed at great length when I made the PR to implement a naming convention in the first place, last year. My actual proposition was for suffixes in modules but prefixes in assemblies, _and_ to name files by subject matter and not a direct copy of the title. The pros are that this makes the file more static in case the module title is slightly tweaked over time, it can keep files names a reasonable and reliable length (some are longer than the actual title, but for long titles, they're always shorter and stay that way), and it makes the TOC chunked by subject matter instead of by exact title wording, which in some cases has articles (An/The, etc) that kind of hinder the alphabetization. And we still make the anchor IDs identical to the file names. This simplifies things tremendously. (+1 @laubai)

Here's what I proposed back then and what our team currently does, for example:

- assembly_guided-decision-tables.adoc (Assembly of guided decision table modules)
- guided-decision-tables-con.adoc (Concept module)
- guided-decision-tables-create-proc.adoc (Procedure module for creating)
- guided-decision-tables-edit-proc.adoc (Procedure module for editing)
- guided-decision-tables-examples-ref.adoc (Reference module with examples)
- guided-decision-tables-limitations-ref.adoc (Reference module with limitations)

For reasons @rkratky mentions, and other reasons, I was outvoted and it was an agree to disagree type thing, which was fine. The compromise, though, was that the mod docs teams said BXMS was welcome to do their own thing with file names if what I proposed worked for us and we strongly agreed on what was ultimately adopted. So that's what we've done. 

But the time to standardize is upon us. I like Sheldon's suggestion, which allows for both types of affix, for now.

**Comment by ssorj** [comment]
I very much agree with what @laubai's view that metadata should not be a required part of the filename.  What's the reason this is considered necessary?  Does some part of the tooling need it?

Asciidoctor has the notion of metadata.  Was that considered and rejected?

https://asciidoctor.org/docs/user-manual/#metadata

**Comment by rkratky** [comment]
@ssorj As I mentioned, including this info in file names isn't for tooling but for writers. Some may not find it helpful, others consider it a big help.

**Comment by rkratky** [comment]
@mishaone 

> that's an interesting perspective. I haven't looked at the templates recently so this may have changed but I would think the content itself would make it clear at a glance what type of module we're in, particularly so if you happen to be a writer familiar with modular docs.

I find it very useful to know without having to open the file. And as @tradej mentioned, it can also help less experienced writers (no guesswork required). Equally useful for me is to have the types grouped.

**Comment by bfallonf** [comment]
@rkratky Sorry, but I don't think any points were addressed. This has been a much more subjective discussion that I was hoping for, because there really is no technical reasons why we need what is essentially (to steal a word from this thread) metadata in a file name.

@tradej Thank you. While I prefer simply the title as the filename, I will be using suffixes. I realise this may change in the future once we have the appropriate tools up and running, so please keep us up-to-date with any changes.

**Comment by ritz303** [comment]
+1 @tradej : Nice proposal!

**Comment by laubai** [comment]
Honestly, I feel like the prefix vs suffix point is just deciding whether to left- or right- mount the door to our bike shed. The key point here for me is "metadata where metadata shouldn't be".

**Comment by mishaone** [comment]
+1 @laubai 

I don't want to keep belaboring the points already made, but I'm not sure any solid reasoning has been presented to keep this information in the file names at all, aside from personal preference, which is fine except when we're deciding a standard for the entire department to follow.

**Comment by tylerkelly13** [comment]
+1 @laubai @mishaone @bfallonf @ssorj 

I was of the understanding that the module type was *required* buy the tooling to be developed here at Red Hat, not just the *preference* of a few.

We should not be adding fluff/complications/bureaucracy, no matter how minor, to how we work internally and with our upstream communities. 

Logically, module file names should reflect only the title and the anchor should also match.

If some product teams can agree internally and with their upstream community that they want module type in the name, then they should be allowed to use prefixes or suffixes, but they should be mandated for every other project. 

The prefix or suffix should be optional, and the modular docs guidelines should reflect this.

**Comment by bfallonf** [comment]
@tylerkelly13 I agree.

Also, I'll bring it up again because I don't see it addressed, but if the modules are titled correctly, then it should be obvious what kind of module it is:

https://redhat-documentation.github.io/modular-docs/#anchor-and-file-names

There's no need for prefixes or suffixes.

**Comment by vikram-redhat** [comment]
@tylerkelly13 - there is no requirement from tooling to know the type of the module. The type of a module is an abstract concept that doesn't drive any tooling decisions, _at this stage_.

**Comment by VladimirSlavik** [comment]
+1 Laura, Tyler, Misha.

The proliferation of prescriptive rules combined with having them change does not work that well "down in the trenches". Having to re-coordinate people after things change makes it a communication and social problem more than a technical one.

... a provocative though. Might the committee imagine _giving up making some of the decisions_? That's an option too.

**Comment by ncbaratta** [comment]
I'd just like to throw in a caution here about any rules made specifically for tooling.  Please remember that many of us work in upstream communities on our documentation and that means we're not the sole decision makers.  No community is going to accept "Because my Red Hat specific tools require it" as a reason to make a decision.  

I've talked with the tools team about this and they're on board.  

Personally, I like having the prefix on the filename and in the communities where I have the power to make that decision I have, but I just want to be sure that everyone is thinking outside of just their own Red Hat repositories.

**Comment by kalexand-rh** [comment]
+1 to not using prefixes or suffixes in the filenames. docs.openshift.com exposes the file name in the URL, so we get links like this one:
https://docs.openshift.com/container-platform/3.9/admin_guide/assembly_replace-master-host.html#manually-adding-etcd-host_replace-master-host

**Comment by JStickler** [comment]
In almost 17 years of writing topic based documentation, I’ve never found the need to include the content type in my file names.  And the few times I’ve experimented with concept/task/reference prefixes in the file names, I’ve always ended up ripping it out because it did not make my life any easier.

If it’s not required by the tooling, I can’t see any benefit to including prefixes OR suffixes in the file name.  Especially for teams that are working with their upstream communities.

**Comment by mishaone** [comment]
There's a surprising amount of talk about personal preferences either way. Are we meant to decide this by a majority vote or based on whether there is any reason besides preference to add this extra requirement? I would assume the latter because of the large impact but I'm still not sure there has been any convincing argument presented to keep this information in file names.

**Comment by vikram-redhat** [comment]
To bring this to a close, I would like to suggest the following:

- There is no standard for file names. Why? Because a) there is enough evidence to suggest that having a meta information in a file name is a personal preference and not something required by tooling b) We can't dictate this standard for upstream communities anyways and c) If indeed it is required by the tooling team at a future date, then we can always add that as an attribute within the files themselves.

- Let's add this (or something similar) **_explicitly_** to the mod docs reference guide: 

> "Assembly and module file names should accurately and closely reflect the title of the assembly or module"

 and take out the existing file name suggestions that require a prefix or suffix.

I can submit a PR to do this and continue the discussion there.

**Comment by ritz303** [comment]
@vikram-redhat : ACK.  I think your suggestions are reasonable and the best way to move this forward.  Please close this issue, and submit your PR.

**Comment by tradej** [comment]
Not that anyone asked, but from a purely process standpoint, I prefer filing a PR that contains the words "Resolves #25", because when that PR is merged, this issue will be closed automatically. Additionally, unless an issue is fixed, it should remain open IMO.

**Comment by sterobin** [comment]
@vikram-redhat, not that I matter, :) but sounds good to me as well. And +1 @tradej.

**Comment by vikram-redhat** [comment]
Thanks all - Added #61 . 

PTAL. 

Once this is accepted, we should send out a notification to the CCS mailing list about this change.

**Comment by mishaone** [comment]
For the people interested in this discussion, there is a bit more happening in the PR linked in the earlier comment.
